Benchmarks can't be compiled on older platforms

When I try to compile the benchmark executable I get the following error from running `make meson`:
```
...
FAILED: benchexe
c++  -o benchexe 'benchexe@exe/benchmarks_main.cpp.o' -Wl,--as-needed -Wl,--no-undefined -Wl,--whole-archive -Wl,--start-group benchmarks/libbench_qsort.a utils/libcpuinfo.a -Wl,--no-whole-archive /usr/local/lib/libbenchmark.a -Wl,--end-group
/usr/bin/ld: /usr/local/lib/libbenchmark.a(benchmark_runner.cc.o): in function `benchmark::internal::BenchmarkRunner::DoNIterations()':
benchmark_runner.cc:(.text+0x1418): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
...
```

This is with v1.8.0 of Google Benchmarks, v1.13.0 of Google's Test Framework. My system is running Ubuntu 20.04 and I have tried both g++-10 and g++-9. Further, the Makefile spits out many errors because my version of GCC can't compile the fp16 files.

I eventually pinned the meson issue down to the fact that my system is running glibc 2.31, where the [linker still needs](https://developers.redhat.com/articles/2021/12/17/why-glibc-234-removed-libpthread) to be explicitly given `-pthread` or `-lpthread` to use threading. This flag is present by default for `gtest` and`gtest_main`, but [Google Benchmark](https://github.com/google/benchmark/pull/1602)'s `pkg-config` will only expose it when you indicate that it's being linked statically. 

The default installation method for both Google's benchmark and testing suite will give a static library, so we _should_ probably be explicitly configuring it as such in the build scripts. As for the Makefile, we could adapt this solution from [Stack Overflow](https://stackoverflow.com/a/5188472) to detect when GCC is not new enough for the `-march=sapphirerapids` flag and exclude the fp16 files.

--- a/.github/workflows/c-cpp.yml
+++ b/.github/workflows/c-cpp.yml
@@ -24,7 +24,7 @@ jobs:
         git clone https://github.com/google/benchmark.git
         cd benchmark
         cmake -E make_directory "build"
-        cmake -E chdir "build" cmake -DBENCHMARK_DOWNLOAD_DEPENDENCIES=on -DCMAKE_BUILD_TYPE=Release ../
+        cmake -E chdir "build" cmake -DBENCHMARK_ENABLE_GTEST_TESTS=OFF -DBENCHMARK_ENABLE_TESTING=OFF -DCMAKE_BUILD_TYPE=Release ../
         sudo cmake --build "build" --config Release --target install
 
     - name: Install Intel SDE
@@ -86,7 +86,7 @@ jobs:
         cd benchmark
         pip3 install -r requirements.txt
         cmake -E make_directory "build"
-        cmake -E chdir "build" cmake -DBENCHMARK_DOWNLOAD_DEPENDENCIES=on -DCMAKE_BUILD_TYPE=Release ../
+        cmake -E chdir "build" cmake -DBENCHMARK_ENABLE_GTEST_TESTS=OFF -DBENCHMARK_ENABLE_TESTING=OFF -DCMAKE_BUILD_TYPE=Release ../
         sudo cmake --build "build" --config Release --target install
 
     - name: Run bench-compare

--- a/.gitignore
+++ b/.gitignore
@@ -33,5 +33,6 @@
 *.out
 *.app
 
+# Build or IDE artifacts
 **/.vscode
-
+/builddir/

--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,11 @@
 CXX		?= g++-12
+CXXFLAGS	+= -I$(SRCDIR) -I$(UTILS) -O3
+GTESTCFLAGS	= `pkg-config --cflags gtest_main`
+GTESTLDFLAGS	= `pkg-config --static --libs gtest_main`
+GBENCHCFLAGS	= `pkg-config --cflags benchmark`
+GBENCHLDFLAGS	= `pkg-config --static --libs benchmark`
+MARCHFLAG	= -march=sapphirerapids
+
 SRCDIR		= ./src
 TESTDIR		= ./tests
 BENCHDIR	= ./benchmarks
@@ -8,13 +15,16 @@ TESTS		= $(wildcard $(TESTDIR)/*.cpp)
 BENCHS		= $(wildcard $(BENCHDIR)/*.cpp)
 TESTOBJS	= $(patsubst $(TESTDIR)/%.cpp,$(TESTDIR)/%.o,$(TESTS))
 BENCHOBJS	= $(patsubst $(BENCHDIR)/%.cpp,$(BENCHDIR)/%.o,$(BENCHS))
-BENCHOBJS	:= $(filter-out $(BENCHDIR)/main.o ,$(BENCHOBJS))
-CXXFLAGS	+= -I$(SRCDIR) -I$(UTILS)
-GTESTCFLAGS	= `pkg-config --cflags gtest_main`
-GTESTLDFLAGS	= `pkg-config --libs gtest_main`
-GBENCHCFLAGS	= `pkg-config --cflags benchmark`
-GBENCHLDFLAGS	= `pkg-config --libs benchmark`
-MARCHFLAG	= -march=sapphirerapids -O3
+
+# Compiling AVX512-FP16 instructions isn't possible for g++ < 12
+ifeq ($(shell expr `$(CXX) -dumpversion | cut -d '.' -f 1` \< 12), 1)
+	MARCHFLAG = -march=icelake-client
+	BENCHOBJS_SKIP += bench-qsortfp16.o
+	TESTOBJS_SKIP += test-qsortfp16.o
+endif
+
+BENCHOBJS	:= $(filter-out $(addprefix $(BENCHDIR)/, $(BENCHOBJS_SKIP)) ,$(BENCHOBJS))
+TESTOBJS	:= $(filter-out $(addprefix $(TESTDIR)/, $(TESTOBJS_SKIP)) ,$(TESTOBJS))
 
 all : test bench
 
@@ -31,7 +41,7 @@ $(BENCHDIR)/%.o : $(BENCHDIR)/%.cpp $(SRCS)
 	$(CXX) $(CXXFLAGS) $(MARCHFLAG) $(GBENCHCFLAGS) -c $< -o $@
 
 bench: $(BENCHOBJS) $(UTILS)/cpuinfo.o
-	$(CXX) $(BENCHDIR)/main.cpp $(BENCHOBJS) $(MARCHFLAG) $(CXXFLAGS) $(GBENCHLDFLAGS) $(UTILS)/cpuinfo.o -o benchexe
+	$(CXX) $(BENCHOBJS) $(UTILS)/cpuinfo.o $(MARCHFLAG) $(CXXFLAGS) -lbenchmark_main $(GBENCHLDFLAGS) -o benchexe
 
 meson:
 	meson setup --warnlevel 0 --buildtype plain builddir

--- a/benchmarks/main.cpp
+++ b/benchmarks/main.cpp
@@ -1,3 +1,0 @@
-#include <benchmark/benchmark.h>
-
-BENCHMARK_MAIN();

--- a/meson.build
+++ b/meson.build
@@ -6,8 +6,8 @@ src = include_directories('src')
 bench = include_directories('benchmarks')
 utils = include_directories('utils')
 tests = include_directories('tests')
-gtest_dep = dependency('gtest_main', required : true)
-gbench_dep = dependency('benchmark', required : true)
+gtest_dep = dependency('gtest_main', required : true, static: true)
+gbench_dep = dependency('benchmark', required : true, static: true)
 
 fp16code = '''#include<immintrin.h>
 int main() { 
@@ -28,9 +28,10 @@ testexe = executable('testexe',
                      link_whole : [libtests, libcpuinfo]
                     )
 
-benchexe = executable('benchexe', 'benchmarks/main.cpp',
+benchexe = executable('benchexe',
                       include_directories : [src, utils, bench],
                       dependencies : [gbench_dep],
+                      link_args: ['-lbenchmark_main'],
                       link_whole : [libbench, libcpuinfo],
                      )
 

